### PR TITLE
mgr/cephadm: add support for nfs with haproxy protocol

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Fixes: https://tracker.ceph.com/issues/58933

Depends on: #47199

This PR adds support for using the HAProxy proxy protocol (v2) with recent builds of nfs ganesha. It builds upon the work in #47199, reusing the new `--ingress-mode` option allowing the user to select between `haproxy-standard` (current behavior) and `haproxy-protocol`. Use of the haproxy protocol will allow nfs ganesha to perform client address based operations behind haproxy.

Still TODO:
* <s>Documentation</s>
* <s>Teuthology test</s>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Parent PR has been merged. Can move out of draft now.

**Comment by phlogistonjohn** [comment]
@adk3798 I moved this from draft to reviewable. I think it would probably be OK for your next teuthology run too.

File: doc/cephadm/services/nfs.rst
Comment (gouthampacha) [inline doc/cephadm/services/nfs.rst:170]: v5

https://github.com/nfs-ganesha/nfs-ganesha/wiki/ReleaseNotes_5

File: src/pybind/mgr/cephadm/tests/test_services.py
Comment (gouthampacha) [inline src/pybind/mgr/cephadm/tests/test_services.py:1294]: Locally, this instruction seems to fail for me with not much in the logs besides: `Cannot chroot /var/lib/haproxy`. When I comment this out, i can deploy the ingress service fine - but that doesn't seem right... I'm using quay.io/ceph/haproxy:2.3 as the image.

**Review by gouthampacha — COMMENTED** [review]
+1 from me. I was able to test this change with OpenStack Manila and use IP client restrictions on the NFS cluster. I've noted an issue I had while testing; Thanks @phlogistonjohn and @adk3798 

```
# ceph nfs cluster create cephfs --ingress --virtual-ip=192.168.130.21 --ingress-mode=haproxy-protocol

# ceph nfs cluster info
{
  "cephfs": {
    "backend": [
      {
        "hostname": "centosstream-9-large-18.service.local",
        "ip": "192.168.130.20",
        "port": 12049
      }
    ],
    "monitor_port": 9049,
    "port": 2049,
    "virtual_ip": "192.168.130.21"
  }
}

# ceph orch ps | grep -e "nfs" -e "ingress"
haproxy.nfs.cephfs.centosstream-9-large-18.uozzqg     centosstream-9-large-18.service.local  *:2049,9049  error            8m ago   2w    11.1M        -  2.3.17-d1c9119  e85424b0d443  7307b7eaaafb
keepalived.nfs.cephfs.centosstream-9-large-18.vikiwf  centosstream-9-large-18.service.local               running (2w)     8m ago   2w    1380k        -  2.1.5           9f7bdb4a87fd  9840f4155b0e
nfs.cephfs.0.0.centosstream-9-large-18.zjrgzs         centosstream-9-large-18.service.local  *:12049      running (2w)     8m ago   2w     119M        -  5               aaae3014e52b  c49977e9c2d4

# ceph orch ls | grep -e "nfs" -e "ingress"
ingress.nfs.cephfs  192.168.130.21:2049,9049      1/4  8m ago     2w   count:2
nfs.cephfs          ?:12049                       1/1  8m ago     2w   count:1
```

File: doc/cephadm/services/nfs.rst
Comment (phlogistonjohn) [reply doc/cephadm/services/nfs.rst:170]: Hah, thanks for catching that I almost certainly would have forgotten I left that TODO in there!

File: src/pybind/mgr/cephadm/tests/test_services.py
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/tests/test_services.py:1294]: The lines in question that you tagged are part of the unit tests. This should not be directly relevant to the code running on a live cluster. I'm guessing the line in the [haproxy template file](https://github.com/ceph/ceph/pull/50614/files/b331d875d285476f8b8ed269d97002afd4d15598#diff-32a0e1886eacaacda4c6a8277c18af9f13634ef5dd0c8e779bf8a49c9b183ff3R4) is  really what you meant to highlight?
If so, I'm unclear why it would behave differently for you than other cephadm deployments of haproxy... this PR doesn't change that part of the config AFAICT.  Are you deploying the container in a (partially) manual way?

File: doc/cephadm/services/nfs.rst
Comment (phlogistonjohn) [reply doc/cephadm/services/nfs.rst]: Fixed

File: src/pybind/mgr/cephadm/tests/test_services.py
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/tests/test_services.py:1294]: Interesting. I tested with images built using the nfs ganesha rpms built by shaman (https://shaman.ceph.com/builds/nfs-ganesha/) but I don't remember exactly which (I think ceph main tag from https://quay.ceph.io/repository/ceph-ci/ceph - but it long enough ago that I forget if it was for sure.

I did not have the problem you encountered. Let's wait for at least one theuthology run of this current series of patches. If it passes the issue may indeed be related to how your image(s) are built, and filing an issue later would be fine.

**Review by adk3798 — APPROVED** [review]
Wasn't able to get my export mounted on the VIP with the port haproxy had open, but was able to if using the host IP and port for the nfs daemon itself (VIP was 192.168.122.100, host IP for host nfs was on was 192.168.122.61, ingress frontend port was 2049, nfs port was 12049)

```
[root@vm-01 ~]# mount -t nfs 192.168.122.100:/test /mnt/test    <---- hangs
[root@vm-01 ~]# mount -t nfs 192.168.122.100:/test /mnt/test -o port=2049  <---- also hangs
[root@vm-01 ~]# mount -t nfs 192.168.122.61:/test /mnt/test -o port=12049  <---- worked
```

but it's possible my container didn't have the correct haproxy version, or I made a setup mistake. Approving anyhow since the code LGTM and we'll see how it goes with the teuth test.

**Comment by adk3798** [comment]
jenkins retest this please

File: doc/mgr/nfs.rst
Comment (phlogistonjohn) [reply doc/mgr/nfs.rst:126]: Yes. Will fix.

File: doc/cephadm/services/nfs.rst
Comment (phlogistonjohn) [reply doc/cephadm/services/nfs.rst:194]: AFAIK http(s) has it own protocol level support for passing client info to the backend server (referenced in https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt ).
However, there's the possibility that some day other non-http, non-nfs servers could support proxy protocol, but none exist right now.

File: src/pybind/mgr/cephadm/services/nfs.py
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/services/nfs.py:124]: Yeah.... I'm in favor of keeping the debug log, but the TODO should probably be fixed. But I'm not sure how.
Can we get the narrow down the list of hosts here to only those this nfs service got scheduled on?  (actually the hosts the haproxy/ingress services got scheduled on but I think thats the same now)

**Comment by adk3798** [comment]
jenkins test make check

File: qa/suites/orch/cephadm/smoke-roleless/2-services/nfs-haproxy-proto.yaml
Comment (adk3798) [inline qa/suites/orch/cephadm/smoke-roleless/2-services/nfs-haproxy-proto.yaml:34]:
```suggestion
      - echo test > /mnt/happy/testfile
```

given the mount is on `/mnt/happy`

File: src/pybind/mgr/cephadm/services/nfs.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/nfs.py:124]: We'd only know the set of haproxy hosts after they've been deployed unfortunately. Although we have the same issue with the keepalived conf needing the IPs for all the hosts other than the ones its running on that will have haproxy daemons for that ingress service. There we handle it by just grabbing all the currently known daemons in the service and then also adding that set of daemons as dependencies for the keepalive daemon. So if the keepalive is deployed before all haproxy daemons are  deployed (meaning it has an incomplete list in its conf when deployed), once the final haproxy daemons are deployed, it will see that the dependencies have changed and reconfig the keepalive with the full set of host IPs it needs. In this particular case, we also just back out if no haproxy daemons are down yet since we know for sure a reconfig will have to happen anyway. It's not perfect since it still means sometimes we're going to have to deploy the keepalive and then reconfigure it just after, but that's the best we came up with for it so far. From the `keepalived_generate_config` in `mgr/cephadm/services/ingress.py`
```
        daemons = self.mgr.cache.get_daemons_by_service(spec.service_name())

        if not daemons and not spec.keepalive_only:
            raise OrchestratorError(
                f'Failed to generate  keepalived.conf: No daemons deployed for {spec.service_name()}')

        deps = sorted([d.name() for d in daemons if d.daemon_type == 'haproxy'])

        host = daemon_spec.host
        hosts = sorted(list(set([host] + [str(d.hostname) for d in daemons])))
```

So we'd probably need something like that, probably minus the check that backs out if no daemons are down. What will likely happen is it will deploy the nfs, deploy the ingress, then reconfig the nfs after once the ingress is deployed and it now knows the set of hosts.

**Review by adk3798 — APPROVED** [review]
Was able to get the mount working when testing this this time, so I think it probably was just too old of a ganesha version in my container last time. This still LGTM, just made one minor comment on the test I think needs to be addressed and responded to a point about grabbing all the haproxy hosts for the ganesha conf

File: qa/suites/orch/cephadm/smoke-roleless/2-services/nfs-haproxy-proto.yaml
Comment (phlogistonjohn) [reply qa/suites/orch/cephadm/smoke-roleless/2-services/nfs-haproxy-proto.yaml]: Fixed.

File: src/pybind/mgr/cephadm/services/nfs.py
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/services/nfs.py:124]: So, maybe I'm just being lazy but I've made my peace with the way the code is at the moment. It basically restricts NFS ganesha from taking proxy protocol values from hosts that are not managed by cephadm. We could do better by restricting it to only those hosts actually running keepalived but it may not be worth the extra effort. We have accomplished the goal of preventing ganesha from being "fronted" by a an arbitrary host talking the haproxy protocol. So unless we lose control of a cephadm managed node I think we could hold off on doing this until we have a good reason.
At the very least I'll change the comment from a "TODO" to a "NB" and if we decide that additonal complexity is worth it I'd ask to do it in a follow up pr. Hows that sound?

File: doc/mgr/nfs.rst
Comment (phlogistonjohn) [reply doc/mgr/nfs.rst]: Fixed.

File: src/pybind/mgr/cephadm/services/nfs.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/nfs.py:124]: I'm alright with that. I don't think having all the hosts in there would do any damage.

Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/services/nfs.py]: Thanks. I think (hope?) that what we have now is good enough. I've updated the comment to reflect these conclusions. I think we're good to go wrt whatever final rounds of testing or discussion is needed before merging.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-05-28_15:20:34-orch:cephadm-wip-adk-testing-2023-05-27-2231-distro-default-smithi/

reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-05-30_19:08:27-orch:cephadm-wip-adk-testing-2023-05-27-2231-distro-default-smithi/

After reruns, 1 failure:

- timeout deploying jaeger-tracing, known issue - https://tracker.ceph.com/issues/59704

**Pull request merged**
